src/org/javarosa/xform/util/ParserCustomizer.java

```diff
@@ -0,0 +1,17 @@
+package org.javarosa.xform.util;
+
+import org.javarosa.xform.parse.XFormParser;
+
+/**
+ * Provides a hook to allow customizing a parser before its parse method is called, allowing the tests
+ * to use the same {@link org.javarosa.xform.util.XFormUtils#getFormFromInputStream(java.io.InputStream)}
+ * call that OpenDataKit Collect uses.
+ */
+public interface ParserCustomizer {
```

Comment:
Is this strictly necessary? It doesn't look like it's key for the fix and I don't know if I feel comfortable with adding more (accidental) complexity to the design: it adds a new way to do something that we could do before, which can be confusing for users.

I would consider a new abstraction aimed to remove the `onWarning` and `onError` methods from the parser instance, which would be a change in break even the way I see it, but I don't know how we could push the needed changes to Collect and how that would be justified.

Reply:
I found a way! :-)

Reply:
I still think there's an issue with this. Honestly, passing an output argument feels worse than the customizer class, especially for a method that's exposed as part of JR's API.

I now realize that the core reason for this change was the need of checking the parse errors and warnings on the tests (and, potentially on clients), which wasn't possible to do with the `XFormParser.onWarning()` and `XFormParser.onError()` methods since the `XFormUtils.getFormFromInputStream()` doesn't give us a chance to subscribe before actually parsing the form.

I think that a workaround for that could be to save those parse messages in the `FormDef` itself. This makes sense to me:
- Clients can't rely 100% on the `onWarning()` and `onError()` methods, as proven by this change we're discussing.
- JR doesn't provide any other means to know if something goes wrong with the parsing of forms (it logs them, but that's not really providing a way for clients to get them).
- It's reasonable that clients want to check/display parse messages after the fact.

I've tried this here: https://github.com/ggalmazor/javarosa/commit/8c6ac1089187df738fb242787090a185abe0d414

Reply:
I like this! I’ll merge in your change. Thanks.